File: TotalCrossVM/src/nm/telephony/SmsManager.c
// Copyright (C) 2000-2013 SuperWaba Ltda.
// Copyright (C) 2014-2020 TotalCross Global Mobile Platform Ltda. 
//
// TotalCross Software Development Kit
//
// This library and virtual machine is distributed in the hope that it will
// be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//
// SPDX-License-Identifier: LGPL-2.1-only

#include "tcvm.h"

#if defined(ANDROID)
#include "android/SmsManager_c.h"
#endif

//////////////////////////////////////////////////////////////////////////
TC_API void ttSM_sendTextMessage_sss(NMParams p) // totalcross/telephony/SmsManager public native void sendTextMessage(String destinationAddress, String scAddress, String text);
{
   TCObject smsManager = p->obj[0];
   TCObject destinationAddress = p->obj[1];
   TCObject scAddress = p->obj[2];
   TCObject text = p->obj[3];
   
   if (destinationAddress == null) {
      throwNullArgumentException(p->currentContext, "destinationAddress");
   } else if (text == null) {
      throwNullArgumentException(p->currentContext, "text");
   } else if (String_charsLen(destinationAddress) == 0) {
      throwIllegalArgumentException(p->currentContext, "Argument destinationAddress cannot be empty");
   } else if (String_charsLen(text) == 0) {
      throwIllegalArgumentException(p->currentContext, "Argument text cannot be empty");
   } else {
#if defined (ANDROID)
      sendTextMessage(destinationAddress, scAddress, text);
#endif
   }
}
//////////////////////////////////////////////////////////////////////////
TC_API void ttSM_registerSmsReceiver_si(NMParams p) // totalcross/telephony/SmsManager public native void registerSmsReceiver(totalcross.telephony.SmsReceiver receiver, int port);
{
   TCObject smsManager = p->obj[0];
   TCObject smsReceiver = p->obj[1];
   int32 port = p->i32[0];
   SmsManager_smsReceiver(smsManager) = smsReceiver;
   
#if defined (ANDROID)   
   registerSmsReceiver(smsReceiver, port);
#endif
}
//////////////////////////////////////////////////////////////////////////
TC_API void ttSM_sendDataMessage_ssiB(NMParams p) // totalcross/telephony/SmsManager public native void sendDataMessage(String destinationAddress, String scAddress, int port, byte[] data);
{
	TCObject smsManager = p->obj[0];
	TCObject destinationAddress = p->obj[1];
	TCObject scAddress = p->obj[2];
	int32 port = p->i32[0];
	TCObject data = p->obj[3];
	
	if (destinationAddress == null) {
		throwNullArgumentException(p->currentContext, "destinationAddress");
	} else if (data == null) {
		throwNullArgumentException(p->currentContext, "data");
	} else if (String_charsLen(destinationAddress) == 0) {
		throwIllegalArgumentException(p->currentContext, "Argument destinationAddress cannot be empty");
	} else if (ARRAYOBJ_LEN(data) == 0) {
		throwIllegalArgumentException(p->currentContext, "Argument data cannot be empty");
	} else {
#if defined (ANDROID)
		sendDataMessage(destinationAddress, scAddress, port, data);
#endif
	}
}
#ifdef ENABLE_TEST_SUITE
//#include "SmsManager_test.h"
#endif

File: TotalCrossSDK/src/main/java/totalcross/telephony/SmsManager.java
// Copyright (C) 2017-2020 TotalCross Global Mobile Platform Ltda.
//
// TotalCross Software Development Kit
//
// This library and virtual machine is distributed in the hope that it will
// be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//
// SPDX-License-Identifier: LGPL-2.1-only

package totalcross.telephony;

import com.totalcross.annotations.ReplacedByNativeOnDeploy;

public class SmsManager {

  private static SmsManager instance;

  private SmsManager() {
  }

  /** Get the default SmsManager. */
  public static SmsManager getDefault() {
    if (instance == null) {
      instance = new SmsManager();
    }
    return instance;
  }

  SmsReceiver receiver;

  /**
   * Registers a receiver to listen for incoming sms messages
   *
   * @param receiver the receiver that will handle incoming sms messages or null to stop listening
   *     incoming messages
   * @param port the port to listen for incoming sms messages - required for data sms messages, pass -1 to listen on default port for text messages
   */
  @ReplacedByNativeOnDeploy
  public void registerSmsReceiver(SmsReceiver receiver, int port) {
    this.receiver = receiver;
  }

  /**
   * Registers a receiver to listen for incoming text sms messages
   *
   * @param receiver the receiver that will handle incoming text sms messages or null to stop listening
   *     incoming messages
   */
  public void registerSmsReceiver(SmsReceiver receiver) {
    this.registerSmsReceiver(receiver, -1);
  }

  /**
   * Send a text based SMS.
   *
   * @param destinationAddress the address to send the message to
   * @param scAddress the service center address or null to use the current default SMSC
   * @param text the body of the message to send
   */
  @ReplacedByNativeOnDeploy
  public void sendTextMessage(String destinationAddress, String scAddress, String text) {
    if (destinationAddress.length() == 0) {
      throw new IllegalArgumentException("Argument destinationAddress cannot be empty");
    }
    if (text.length() == 0) {
      throw new IllegalArgumentException("Argument text cannot be empty");
    }
    System.out.println("To: " + destinationAddress + ", From: " + scAddress + ", Text: " + text);
  }

  /**
   * Send a data based SMS to a specific application port.
   *
   * @param destinationAddress the address to send the message to
   * @param scAddress the service center address or null to use the current default SMSC
   * @param port the port to deliver the message to
   * @param data the body of the message to send
   */
  @ReplacedByNativeOnDeploy
  public void sendDataMessage(String destinationAddress, String scAddress, int port, byte[] data) {
    if (destinationAddress.length() == 0) {
      throw new IllegalArgumentException("Argument destinationAddress cannot be empty");
    }
    if (data.length == 0) {
      throw new IllegalArgumentException("Argument data cannot be empty");
    }
    System.out.println("To: " + destinationAddress + ", From: " + scAddress + ", Data with length: " + data.length);
  }
}
